File: 752-OpenTheLock/BFS.java
import java.util.Queue;
import java.util.Set;

class Solution {
    public int openLock(String[] deadends, String target) {
int step = 0;
        Queue<String> queue = new ArrayDeque<>();
        Set<String> visited = new HashSet<>();
        Set<String> dead = new HashSet<>();
        for (String s : deadends) {
            dead.add(s);
        }
        queue.add("0000");
        visited.add("0000");
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String tmp = queue.poll();
                if (dead.contains(tmp)) {
                    continue;
                }
                if (tmp.equals(target)) {
                    return step;
                }
                for (int j = 0; j < 4; j++) {
                    String up = plusOne(tmp, j);
                    if (!visited.contains(up)) {
                        queue.add(up);
                        visited.add(up);
                    }
                    String down = minusOne(tmp, j);
                    if (!visited.contains(down)) {
                        queue.add(down);
                        visited.add(down);
                    }
                }
            }
            step++;
        }
        return -1;
    }
    private String minusOne(String tmp, int j) {
        char[] ch = tmp.toCharArray();
        if (ch[j] == '0') {
            ch[j] = '9';
        } else {
            ch[j] -= 1;
        }
        return new String(ch);
    }

    private String plusOne(String tmp, int j) {
        char[] ch = tmp.toCharArray();
        if (ch[j] == '9') {
            ch[j] = '0';
        } else {
            ch[j] += 1;
        }
        return new String(ch);
    }
}

File: 752-OpenTheLock/BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>
using namespace std;

class Solution {
public: 
    int openLock(vector<string>& deadends, string target) {
        const string start = "0000";
        int step = 0;
        queue<string> queue;
        unordered_set<string> visited;
        unordered_set<string> dead;
        for (string s : deadends) {
            dead.insert(s);
        }
        queue.push(start);
        while (!queue.empty()) {
            ++step;
            int size = queue.size(); //本轮需要处理的个数
            for (int i = 0; i < size; i++) {
                const string cur = queue.front();
                queue.pop();
                for (int j = 0; j < 4; j++) { //改变每一位
                    for (int k = -1; k <= 1; k += 2) { //加一或减一
                        string temp = cur;
                        temp[j] = (temp[j] -'0' + k + 10) % 10 + '0';
                        if (temp == target) return step;
                        if (dead.count(temp) || visited.count(temp)) continue;
                        queue.push(temp);
                        visited.insert(temp);
                    }
                }
            }
        }
        return -1;
    }
};